test_scripts/clientconf.go

```diff
@@ -0,0 +1,182 @@
+package main
+
+import (
+	"encoding/binary"
+	"encoding/hex"
+	"flag"
+	"fmt"
+	"github.com/golang/protobuf/proto"
+	pb "github.com/sergeyfrolov/gotapdance/protobuf"
+	"io/ioutil"
+	"log"
+	"net"
+)
+
+func printClientConf(clientConf pb.ClientConf) {
+	fmt.Printf("Generation: %d\n", clientConf.GetGeneration())
+	fmt.Printf("Default Pubkey: %s\n", hex.EncodeToString(clientConf.GetDefaultPubkey().Key[:]))
+	decoys := clientConf.DecoyList.TlsDecoys
+	fmt.Printf("Decoy List: %d decoys\n", len(decoys))
+	for i, decoy := range decoys {
+		ip := make(net.IP, 4)
+		binary.BigEndian.PutUint32(ip, decoy.GetIpv4Addr())
+		fmt.Printf("%d:\n  %s (%s)\n", i, decoy.GetHostname(), ip.To4().String())
+		if decoy.GetPubkey() != nil {
+			fmt.Printf("  pubkey: %s\n", hex.EncodeToString(decoy.GetPubkey().Key[:]))
+		}
+		if decoy.GetTimeout() != 0 {
+			fmt.Printf("  timeout: %d ms\n", decoy.GetTimeout())
+		}
+		if decoy.GetTcpwin() != 0 {
+			fmt.Printf("  tcpwin: %d bytes\n", decoy.GetTcpwin())
+		}
+	}
+
+}
+
+func parseClientConf(fname string) pb.ClientConf {
+
+	clientConf := pb.ClientConf{}
+	buf, err := ioutil.ReadFile(fname)
+	if err != nil {
+		log.Fatal("Error reading file:", err)
+	}
+	err = proto.Unmarshal(buf, &clientConf)
+	if err != nil {
+		log.Fatal("Error parsing ClientConf", err)
+	}
+	return clientConf
+}
+
+func parsePubkey(pubkey string) []byte {
+	pubkey_bin, err := hex.DecodeString(pubkey)
+	if err != nil {
+		log.Fatal("Error parsing pubkey:", err)
+	}
+	if len(pubkey_bin) != 32 {
+		log.Fatal("Error: pubkey length: expected 32, got ", len(pubkey_bin))
+	}
+	return pubkey_bin
+
+}
+
+func updateDecoy(decoy *pb.TLSDecoySpec, host string, ip string, pubkey string, delpubkey bool, timeout int, tcpwin int) {
+
+	if host != "" {
+		decoy.Hostname = &host
+	}
+	if ip != "" {
+		ip4 := binary.BigEndian.Uint32(net.ParseIP(ip).To4())
```

Comment:
`net.ParseIP(ip).To4()` could return nil
and this whole thing will panic. Which is actually way better than filling it with 0s or something, but might be worth printing proper error